File: apps/app-frontend/src/helpers/servers.ts
import type { ServerTypeId } from '@modrinth/server'
import { invoke } from '@tauri-apps/api/core'
import { listen } from '@tauri-apps/api/event'

export interface ModpackInfoData {
	projectId: string
	versionId: string
	title: string
	iconUrl?: string
}

export interface ServerManifestData {
	id: string
	name: string
	serverType: ServerTypeId
	gameVersion: string
	loaderVersion?: string
	jarName?: string
	iconPath?: string
	modpack?: ModpackInfoData
	installState?: 'incomplete' | 'failed' | null
	installError?: string | null
	javaPath?: string
	memoryMb?: number
	jvmArgs: string[]
	createdAt: string
	lastStartedAt?: string
	lastExitCrashed: boolean
}

export interface ServerInfoData extends ServerManifestData {
	path: string
	running: boolean
	eulaExists: boolean
	eulaAccepted: boolean
	port: number | null
}

export type ServerExitReason = 'eula'

export type ServerEventPayload =
	| { event: 'log'; line: string }
	| { event: 'download_progress'; downloaded: number; total?: number }
	| { event: 'started' }
	| { event: 'stopped'; crashed: boolean; reason?: ServerExitReason }
	| { event: 'eula_required'; server_id: string; eula_text: string }

export interface PortProcessInfoData {
	pid: number
	name?: string | null
}

export interface InstallModpackOptions {
	mrpackUrl: string
	mrpackSha1?: string
	jarUrl: string
	jarFilename: string
	jarSha1?: string
	javaPath?: string
	modpackProjectId?: string
	modpackVersionId?: string
	modpackTitle?: string
	modpackIconUrl?: string
}

const command = (name: string) => `plugin:servers|${name}`

export const servers = {
	list: () => invoke<ServerInfoData[]>(command('servers_list')),
	get: (serverId: string) => invoke<ServerInfoData>(command('servers_get'), { serverId }),
	create: (options: {
		name: string
		serverType: ServerTypeId
		gameVersion: string
		loaderVersion?: string
		javaPath?: string
		memoryMb?: number
	}) => invoke<ServerManifestData>(command('servers_create'), options),
	updateSettings: (
		serverId: string,
		options: {
			name?: string
			javaPath?: string
			memoryMb?: number
			jvmArgs?: string[]
		},
	) => invoke<ServerManifestData>(command('servers_update_settings'), { serverId, ...options }),
	setIcon: (serverId: string, iconPath: string | null) =>
		invoke<ServerManifestData>(command('servers_set_icon'), { serverId, iconPath }),
	delete: (serverId: string) => invoke<void>(command('servers_delete'), { serverId }),
	readFile: (serverId: string, file: string) =>
		invoke<string>(command('servers_read_file'), { serverId, file }),
	writeFile: (serverId: string, file: string, contents: string) =>
		invoke<void>(command('servers_write_file'), { serverId, file, contents }),
	downloadFile: (serverId: string, url: string, filename: string, expectedSha1?: string) =>
		invoke<void>(command('servers_download_file'), {
			serverId,
			url,
			filename,
			expectedSha1,
		}),
	installModpack: (serverId: string, options: InstallModpackOptions) =>
		invoke<void>(command('servers_install_modpack'), {
			serverId,
			...options,
			javaPath: options.javaPath ?? null,
		}),
	installForge: (serverId: string, mcVersion: string, build: string, javaPath?: string) =>
		invoke<void>(command('servers_install_forge'), { serverId, mcVersion, build, javaPath }),
	start: (
		serverId: string,
		options?: { javaPath?: string; memoryMb?: number; jvmArgs?: string[] },
	) => invoke<void>(command('servers_start'), { serverId, ...options }),
	sendCommand: (serverId: string, commandText: string) =>
		invoke<void>(command('servers_send_command'), { serverId, command: commandText }),
	stop: (serverId: string) => invoke<void>(command('servers_stop'), { serverId }),
	kill: (serverId: string) => invoke<void>(command('servers_kill'), { serverId }),
	killPortProcess: (port: number) => invoke<void>(command('servers_kill_port_process'), { port }),
	portProcess: (port: number) =>
		invoke<PortProcessInfoData | null>(command('servers_port_process'), { port }),
	getLogBuffer: (serverId: string) =>
		invoke<string[]>(command('servers_get_log_buffer'), { serverId }),
	clearLog: (serverId: string) => invoke<void>(command('servers_clear_log'), { serverId }),
}

export async function serverEventListener(
	callback: (serverId: string, payload: ServerEventPayload) => void,
): Promise<() => void> {
	const unlisten = await listen<{ serverId: string; event: string } & ServerEventPayload>(
		'server',
		(event) => callback(event.payload.serverId, event.payload),
	)
	return unlisten
}

File: apps/app/src/api/servers.rs
use crate::api::Result;
use theseus::prelude::*;
use theseus::servers::ServerInfo;

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("servers")
        .invoke_handler(tauri::generate_handler![
            servers_list,
            servers_get,
            servers_create,
            servers_update_settings,
            servers_set_icon,
            servers_delete,
            servers_read_file,
            servers_write_file,
            servers_download_file,
            servers_install_forge,
            servers_install_modpack,
            servers_start,
            servers_send_command,
            servers_stop,
            servers_kill,
            servers_kill_port_process,
            servers_port_process,
            servers_get_log_buffer,
            servers_clear_log,
        ])
        .build()
}

#[tauri::command]
pub async fn servers_list() -> Result<Vec<ServerInfo>> {
    Ok(servers::list().await?)
}

#[tauri::command]
pub async fn servers_get(server_id: &str) -> Result<ServerInfo> {
    Ok(servers::get(server_id).await?)
}

#[tauri::command]
pub async fn servers_create(
    name: &str,
    server_type: &str,
    game_version: &str,
    loader_version: Option<String>,
    java_path: Option<String>,
    memory_mb: Option<u32>,
) -> Result<servers::ServerManifest> {
    Ok(servers::create(
        name,
        server_type,
        game_version,
        loader_version,
        java_path,
        memory_mb,
    )
    .await?)
}

#[tauri::command]
pub async fn servers_update_settings(
    server_id: &str,
    name: Option<String>,
    java_path: Option<String>,
    memory_mb: Option<u32>,
    jvm_args: Option<Vec<String>>,
) -> Result<servers::ServerManifest> {
    Ok(servers::update_settings(
        server_id, name, java_path, memory_mb, jvm_args,
    )
    .await?)
}

#[tauri::command]
pub async fn servers_set_icon(
    server_id: &str,
    icon_path: Option<String>,
) -> Result<servers::ServerManifest> {
    Ok(servers::set_icon(server_id, icon_path).await?)
}

#[tauri::command]
pub async fn servers_delete(server_id: &str) -> Result<()> {
    Ok(servers::delete(server_id).await?)
}

#[tauri::command]
pub async fn servers_read_file(server_id: &str, file: &str) -> Result<String> {
    Ok(servers::read_file(server_id, file).await?)
}

#[tauri::command]
pub async fn servers_write_file(
    server_id: &str,
    file: &str,
    contents: &str,
) -> Result<()> {
    Ok(servers::write_file(server_id, file, contents).await?)
}

#[tauri::command]
pub async fn servers_download_file(
    server_id: &str,
    url: &str,
    filename: &str,
    expected_sha1: Option<String>,
) -> Result<()> {
    Ok(servers::download_file(server_id, url, filename, expected_sha1).await?)
}

#[tauri::command]
pub async fn servers_install_forge(
    server_id: &str,
    mc_version: &str,
    build: &str,
    java_path: Option<String>,
) -> Result<()> {
    Ok(servers::install_forge(server_id, mc_version, build, java_path).await?)
}

#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub async fn servers_install_modpack(
    server_id: &str,
    mrpack_url: &str,
    mrpack_sha1: Option<String>,
    jar_url: &str,
    jar_filename: &str,
    jar_sha1: Option<String>,
    java_path: Option<String>,
    modpack_project_id: Option<String>,
    modpack_version_id: Option<String>,
    modpack_title: Option<String>,
    modpack_icon_url: Option<String>,
) -> Result<()> {
    Ok(servers::install_modpack(
        server_id,
        mrpack_url,
        mrpack_sha1,
        jar_url,
        jar_filename,
        jar_sha1,
        java_path,
        modpack_project_id,
        modpack_version_id,
        modpack_title,
        modpack_icon_url,
    )
    .await?)
}

#[tauri::command]
pub async fn servers_start(
    server_id: &str,
    java_path: Option<String>,
    memory_mb: Option<u32>,
    jvm_args: Option<Vec<String>>,
) -> Result<()> {
    Ok(servers::start(server_id, java_path, memory_mb, jvm_args).await?)
}

#[tauri::command]
pub async fn servers_send_command(
    server_id: &str,
    command: &str,
) -> Result<()> {
    Ok(servers::send_command(server_id, command).await?)
}

#[tauri::command]
pub async fn servers_stop(server_id: &str) -> Result<()> {
    Ok(servers::stop(server_id).await?)
}

#[tauri::command]
pub async fn servers_kill(server_id: &str) -> Result<()> {
    Ok(servers::kill(server_id).await?)
}

#[tauri::command]
pub async fn servers_kill_port_process(port: u16) -> Result<()> {
    Ok(servers::kill_port_process(port).await?)
}

#[tauri::command]
pub async fn servers_port_process(
    port: u16,
) -> Result<Option<servers::PortProcessInfo>> {
    Ok(servers::port_process(port).await?)
}

#[tauri::command]
pub async fn servers_get_log_buffer(server_id: &str) -> Result<Vec<String>> {
    Ok(servers::get_log_buffer(server_id).await?)
}

#[tauri::command]
pub async fn servers_clear_log(server_id: &str) -> Result<()> {
    Ok(servers::clear_log(server_id).await?)
}
